client/src/context/useChatContext.tsx

```diff
@@ -155,35 +165,33 @@ export const ChatProvider: FunctionComponent = ({ children }) => {
 				});
 			}
 		},
-		[removeChatMessages, updateSnackBarMessage]
+		[removeChatMessages, updateSnackBarMessage, loggedInUser]
 	);
 
-	const handleNewMessage = useCallback(
+	// create a new userChat when received via socket. If you created the chat (see above function) you will set that chat as the active chat.
+	useEffect(() => {
+		if (newUserChat) {
+			setUserChats((state) => {
+				if (!state) return [newUserChat];
+				// new chat will appear at top of list
+				return [newUserChat, ...state];
+			});
+			resetNewUserChat();
+		}
+	}, [newUserChat, resetNewUserChat]);
```

Comment:
It's an interesting choice to do this in a `useEffect`; is there any reason you can't do this logic in the socket event handler directly? Keep in mind that this causes extra component re-renders since you're updating state to handle the socket event, then updating state again in response to the initial state change.

Reply:
I don't believe i can do this from the socket event handler as this 'ChatProvider" is the holder of userChats state. 

As the SocketProvider wraps around the ChatProvider the SocketProvider can't update the userChat state as that would be providing state upward in the tree. When the socket advises that there is a new chat the user needs to join it advises the ChatProvider to update it's state. 

This was created for the use case of when two users are logged in (userA & userB) if userA creates a new chat with userB. Once the sockets are connected to the chat room (via the SocketProvider) they then tell the ChatProvider about the chat room they joined so that the ChatProvided can add it into userChat state.

I would love your thoughts on this as it is cause more renders that i would like.

Reply:
There are a few ways to refactor this to improve performance. One of the simplest ways is to remove the context from the socket logic and just keep the state and event handlers in a custom hook, then pass callback functions as arguments to the hook that get called when an event is received.

Just as an example:

```
useSocket({
    handleNewMessage: (socketData) => {
        // add event data from socket to chat state
    },
    handleNewChat: (socketData) => {
        // add event data from socket to chat state
    }
});
```

Reply:
Great i like this concept